File: src/search/scala/BinarySearch.scala
package search.scala

import sort.scala.{HeapSort}
import scala.annotation.tailrec
object BinarySearch extends App {


  /**
    * time complexity O(log n)
    * space complexity O(1) in place search algorithm
    * @param arr
    * @param key
    * @return index of the key
    */
  def search(arr: Array[Int], key : Int) : Int = {
    @tailrec
    def binarySearch(minIndex: Int, maxIndex: Int): Int = {
      val midIndex = (minIndex + maxIndex) / 2
      if (midIndex > maxIndex)
        -1
      else if (arr(midIndex) == key)
        midIndex
      else if (arr(midIndex) > key)
        binarySearch(minIndex, midIndex -1)
      else
        binarySearch(midIndex + 1, maxIndex)
    }

    binarySearch(0, arr.length - 1)
  }
  val array : Array[Int] = Array(1,3,5,2,6,4,9,0,30,12)
  HeapSort.sort(array)
  array foreach println
  val index = search(arr = array, key = 0)
  index match {
    case -1 => println("key not found")
    case _ => println(s"key found at position $index")
  }
}

File: src/search/java/BinarySearch.java
package search.java;

import sort.java.BubbleSort;

import java.util.Arrays;

public class BinarySearch {

    public static int search(int[] array, int key) {
        return  binarySearch(array, 0, array.length -1, key);
    }

    private static int binarySearch(int arr[], int l, int r, int key) {
        if (r >= 1) {
            int mid = (r + l)/2;
            if(arr[mid] == key) {

            }
        }
        return -1;
    }

    public static void main(String... args) {
        int[] a = {7,8,9,1,2,3,10,23,0};
        BubbleSort.bubbleSort(a);
        int key = 11;
        Arrays.stream(a).forEach(System.out::println);
        System.out.println(String.format("found key : %d", search(a, key)));
    }
}
